nodeup/pkg/model/kubelet.go

```diff
@@ -244,6 +246,12 @@ func (b *KubeletBuilder) buildSystemdEnvironmentFile(kubeletConfig *kops.Kubelet
 		flags += " --tls-private-key-file=" + b.PathSrvKubernetes() + "/kubelet-server.key"
 	}
 
+	if b.Cluster.Spec.IsIPv6Only() {
+		flags += " --address=::"
+		// This makes kubelet prefer ipv6. E.g Pods with host network will have ipv6 addresses.
+		flags += " --node-ip=::"
```

Comment:
This doesn't work when `clould-provider=external` and the address I think is set somewhere else in model.
https://github.com/kubernetes/kubernetes/blob/v1.22.1/pkg/kubelet/nodestatus/setters.go#L95

Reply:
It doesn't affect NodeIP assignment, but this does still affect Pods with HostNetwork. Since we are in a hybrid dual-stack/single-stack mode, these get ipv4 addresses without this, ipv6 addresses with.

Reply:
How? There is no code for taking this into consideration with external CCM?

Reply:
This only depends on the oder the CCM sends the IP addresses, unless something changed since I tested.

Reply:
I have largely tested this on 1.21, so I doubt it changed. But there may also be different behavior for static Pods and the timing of when CCM sets the Node IPs. Node objects will exist before CCM modifies them.

I did find that I needed both this flag _and_ the ordering of internal addresses to get the addresses correctly.

Worst case, this one is noop though, so there shouldn't be any harm in keeping it.

Reply:
It is a harm because of the confusion it creates and expectations. 1.21 at least with kOps still defaults to KCM.
I suggest to test with 1.22 and have some scenarios that we can also reproduce.

I think what you are observing is somehow correct. The IPv6 only behaviour for dual-stack nodes is not exactly great and some things that should be done by kubelet are not quite well implemented and documented.

If we want to go this way, which is not exactly ideal, we may as well set the NodeIP to the first IPv6 address from the EC2 metadata.

Reply:
Makes sense.

I removed the commit. The important bit was getting IPs into the kubelet certs.